### PR TITLE
Implement CertifyKey with CSRs and add encodings of CMS/CSRs

### LINKED ISSUE #83: Implement DPE Command CertifyCsr
See https://github.com/TrustedComputingGroup/Server-Internal/blob/main/dpe-irot-profile/main.md#certifycsr for specification.

### LINKED ISSUE #237: build.rs is not conditionally compiling correctly
Since build.rs runs before compilation, it doesn't know about features. Thus, The default value of dpe_profile_p256_sha256 is always getting used. This causes issues with the openssl implementation of ecdsa_sign_with_alias since we end up always signing with a 32 bit private key.

### PR DESCRIPTION
**Pull request by sree-revoori1**
Also, rework build.rs to write both the 256 bit and 384 bit private keys. This is necessary because feature compilation doesn't work in build.rs. Then when we read the alias key in OpensslCrypto::ecdsa_sign_with_alias, we can read from the proper file based on the value of the algorithm size. 

fixes #237 
fixes #83

### CODE REVIEW

File: dpe/src/x509.rs [hunk 1]
```diff
@@ -1030,6 +1160,149 @@ impl X509CertWriter<'_> {
...
 
+    /// Encode a SignedData
+    ///
+    /// SignedData ::= SEQUENCE {
+    /// version CMSVersion,
+    /// digestAlgorithms DigestAlgorithmIdentifiers,
+    /// encapContentInfo EncapsulatedContentInfo,
+    /// certificates [0] IMPLICIT CertificateSet OPTIONAL,
+    /// crls [1] IMPLICIT RevocationInfoChoices OPTIONAL,
+    /// signerInfos SignerInfos }
```
Comment (jhand2) [inline dpe/src/x509.rs:1171]: minor nit: Can you indent these fields?

File: dpe/src/x509.rs [hunk 2]
```diff
@@ -1030,6 +1160,149 @@ impl X509CertWriter<'_> {
         Ok(bytes_written)
     }
 
+    /// Encode a SignedData
```
Comment (jhand2) [inline dpe/src/x509.rs:1163]: nit: Might be good to add a comment that we don't populate the `certificates` or `crls` fields

File: dpe/src/x509.rs [hunk 3]
```diff
@@ -1030,6 +1160,149 @@ impl X509CertWriter<'_> {
...
+
+        // signerInfos
+        bytes_written += self.encode_signer_infos(serial_number, issuer_name, sig)?;
+
+        Ok(bytes_written)
+    }
+
+    /// Encode a SET OF SignerInfo
+    ///
+    /// SignerInfo ::= SEQUENCE {
```
Comment (jhand2) [inline dpe/src/x509.rs:1211]: nit: same comment about indentation

File: dpe/src/commands/certify_key.rs [hunk 4]
```diff
@@ -150,7 +150,58 @@ impl CommandExecution for CertifyKeyCmd {
...
+                    .crypto
+                    .hash(
+                        DPE_PROFILE.alg_len(),
+                        &cert_req_info_buffer[..bytes_written],
+                    )
+                    .map_err(|_| DpeErrorCode::HashError)?;
+                let cert_req_info_sig = env
+                    .crypto
+                    .ecdsa_sign_with_alias(DPE_PROFILE.alg_len(), &cert_req_info_digest)
+                    .map_err(|_| DpeErrorCode::CryptoError)?;
```
Comment (jhand2) [inline dpe/src/commands/certify_key.rs:178]: PKCS#10 CSRs should be self-signed (signed the the private key returned by `derive_key_pair`). This proves ownership of the key (this is also why we wrap the CSR in a CMS structure. So we can sign the whole thing with a different key).

File: dpe/src/x509.rs [hunk 5]
```diff
@@ -1135,6 +1408,118 @@ impl X509CertWriter<'_> {
...
+
+        Ok(bytes_written)
+    }
+
+    /// Encode a CMS ContentInfo message
+    ///
+    /// ContentInfo ::= SEQUENCE {
+    /// contentType ContentType,
+    /// content [0] EXPLICIT ANY DEFINED BY contentType }
+    pub fn encode_cms(
```
Comment (jhand2) [inline dpe/src/x509.rs:1497]: Can you add some unit tests to make sure the certs can be parsed?

The certification_request module from x509_parser should work for the PKCS#10 stuff: https://docs.rs/x509-parser/latest/x509_parser/certification_request/index.html

And then the CMS crate you added for parsing the outer CMS part.

File: dpe/src/x509.rs [hunk 6]
```diff
@@ -1135,6 +1408,118 @@ impl X509CertWriter<'_> {
...
+        bytes_written += self.encode_version()?;
+
+        // subject
+        bytes_written += self.encode_rdn(subject_name)?;
+
+        // subjectPublicKeyInfo
+        bytes_written += self.encode_ecdsa_subject_pubkey_info(pub_key)?;
+
+        // extensions
+        bytes_written += self.encode_extensions(measurements)?;
```
Comment (jhand2) [inline dpe/src/x509.rs:1454]: I think there needs to be a wrapper struct around the extensions. There is an attribute "Extension Request" defined in RFC 2985 that should wrap these extensions.

File: dpe/src/commands/certify_key.rs (hunk 4 above)
Comment (sree-revoori1) [reply dpe/src/commands/certify_key.rs:178]: Done

File: dpe/src/x509.rs (hunk 2 above)
Comment (sree-revoori1) [reply dpe/src/x509.rs:1163]: Done

File: dpe/src/x509.rs (hunk 1 above)
Comment (sree-revoori1) [reply dpe/src/x509.rs:1171]: Done

File: dpe/src/x509.rs (hunk 3 above)
Comment (sree-revoori1) [reply dpe/src/x509.rs:1211]: Done

File: dpe/src/x509.rs (hunk 6 above)
Comment (sree-revoori1) [reply dpe/src/x509.rs]: Done

File: dpe/src/x509.rs (hunk 5 above)
Comment (sree-revoori1) [reply dpe/src/x509.rs:1683]: Done

File: dpe/src/commands/certify_key.rs [hunk 7]
```diff
@@ -329,4 +402,214 @@ mod tests {
...
+        };
+        env.crypto
+            .get_pubkey_serial(DPE_PROFILE.alg_len(), &pub_key, &mut subj_serial)
+            .unwrap();
+        match sid {
+            SignerIdentifier::IssuerAndSerialNumber(issuer_and_serial_number) => {
+                let cert_serial_number = &issuer_and_serial_number.serial_number;
+                assert_eq!(&subj_serial[..20], cert_serial_number.as_bytes());
+                // note: the cms crate does not let us convert issuer_and_serial_number.issuer
+                // into bytes so we unfortunately cannot test the issuer name here.
```
Comment (jhand2) [inline dpe/src/commands/certify_key.rs:506]: I don't quite understand. Isn't there an as_bytes function here? https://docs.rs/x509-cert/0.2.3/x509_cert/serial_number/struct.SerialNumber.html#method.as_bytes

File: dpe/src/commands/certify_key.rs [hunk 8]
```diff
@@ -329,4 +402,214 @@ mod tests {
...
+        let pub_key_der = ec_point.data();
+        // skip first 0x04 der encoding byte
+        let x = BigNum::from_slice(&pub_key_der[1..DPE_PROFILE.get_ecc_int_size() + 1]).unwrap();
+        let y = BigNum::from_slice(&pub_key_der[DPE_PROFILE.get_ecc_int_size() + 1..]).unwrap();
+        EcKey::from_public_key_affine_coordinates(group, &x, &y).unwrap();
+
+        // note: ideally here, we would verify the signature with the public key.
+        // However, the x509_parser library does not expose a function to
+        // convert the certification request info into bytes so we have no way to
+        // get the digest.
```
Comment (jhand2) [inline dpe/src/commands/certify_key.rs:565]: I think you can use the `raw` field here: https://docs.rs/x509-parser/latest/x509_parser/certification_request/struct.X509CertificationRequestInfo.html

File: dpe/src/x509.rs [hunk 9]
```diff
@@ -1088,6 +1299,251 @@ impl X509CertWriter<'_> {
...
+    ///
+    /// Attributes ::= SET OF Attribute
+    ///
+    /// Attribute ::= SEQUENCE {
+    ///    attrType OBJECT IDENTIFIER,
+    ///    attrValues SET OF AttributeValue
+    /// }
+    ///
+    /// AttributeValue ::= ANY -- Defined by attribute type
+    #[allow(clippy::identity_op)]
```
Comment (jhand2) [inline dpe/src/x509.rs:1401]: Why is this needed? Because of the `| 0x0`?

File: dpe/src/x509.rs [hunk 10]
```diff
@@ -1088,6 +1299,251 @@ impl X509CertWriter<'_> {
         Ok(bytes_written)
     }
 
+    fn encode_implicit_extensions(
```
Comment (jhand2) [inline dpe/src/x509.rs:1302]: I wonder if we can avoid duplicating the code with a parameter here. Something like

```rust
enum ConstructedTag {
    Implicit(u8),
    Explicit(u8),
}

fn encode_implicit_extensions(
    &mut self,
    measurements: &MeasurementData,
    constructed: Option<ConstructedTag>,
) -> Result<usize, DpeErrorCode> {
    match constructed {
        Some(Implicit(tag_num)) => // Do implicit tag
        Some(Explicit(tag_num)) => // Do explicit tag
        None => // Use normal tag
    }
}
```

File: dpe/src/x509.rs (hunk 9 above)
Comment (sree-revoori1) [reply dpe/src/x509.rs:1401]: Yes

File: dpe/src/commands/certify_key.rs (hunk 7 above)
Comment (sree-revoori1) [reply dpe/src/commands/certify_key.rs:506]: I am using the cms crate to parse the wrapping ContentInfo  - https://docs.rs/cms/0.2.2/cms/cert/struct.IssuerAndSerialNumber.html

Comment (sree-revoori1) [reply dpe/src/commands/certify_key.rs:506]: i was able to use Name::to_der to parse it actually.

File: dpe/src/commands/certify_key.rs (hunk 8 above)
Comment (sree-revoori1) [reply dpe/src/commands/certify_key.rs:565]: Thanks, this worked.

File: dpe/src/x509.rs (hunk 10 above)
Comment (sree-revoori1) [reply dpe/src/x509.rs]: Done, except I just passed a bool to encode_extensions since there's only two cases.

**Review by jhand2 — APPROVED** [review]
